vperfetto-min.cpp

```diff
@@ -0,0 +1,148 @@
+#include "perfetto.h"
+#include "vperfetto-min.h"
+#include "vperfetto.h"
+
+#include <string>
+#include <thread>
+#include <fstream>
+#include <unordered_map>
+
+PERFETTO_DEFINE_CATEGORIES(
+    ::perfetto::Category("gfx")
+        .SetDescription("Events from the graphics subsystem"));
+PERFETTO_TRACK_EVENT_STATIC_STORAGE();
+
+#define TRACE_COUNTER(category, name, value)  \
+    PERFETTO_INTERNAL_TRACK_EVENT(                  \
+            category, name, \
+            ::perfetto::protos::pbzero::TrackEvent::TYPE_COUNTER, [&](::perfetto::EventContext ctx){ \
+            ctx.event()->set_counter_value(value); \
+            })
+
+#ifdef __cplusplus
+#   define CC_LIKELY( exp )    (__builtin_expect( !!(exp), true ))
+#   define CC_UNLIKELY( exp )  (__builtin_expect( !!(exp), false ))
+#else
+#   define CC_LIKELY( exp )    (__builtin_expect( !!(exp), 1 ))
+#   define CC_UNLIKELY( exp )  (__builtin_expect( !!(exp), 0 ))
+#endif
+
+namespace vperfetto {
+
+static bool sPerfettoInitialized = false;
+
+static VirtualDeviceTraceConfig sTraceConfig = {
+    .initialized = false,
+    .tracingDisabled = true,
+    .packetsWritten = 0,
+    .sequenceIdWritten = 0,
+    .currentInterningId = 1,
+    .currentThreadId = 1,
+    .hostFilename = "vmm.trace",
+    .guestFilename = nullptr,
+    .combinedFilename = nullptr,
+    .hostStartTime = 0,
+    .guestStartTime = 0,
+    .guestTimeDiff = 0,
+    .perThreadStorageMb = 1,
+};
+
+struct TraceProgress {
+    std::vector<char> hostTrace;
+    std::vector<char> guestTrace;
+    std::vector<char> combinedTrace;
+};
+
+static TraceProgress sTraceProgress;
+
+static std::unique_ptr<::perfetto::TracingSession> sTracingSession;
+
+static void initPerfetto() {
+    if (!sPerfettoInitialized) {
+        ::perfetto::TracingInitArgs args;
+        args.backends |= ::perfetto::kInProcessBackend;
```

Comment:
I think I will need an option to run Perfetto in system mode, so we also have GPU counter events from an external data source (gfx-pps). Otherwise, we would need for Crosvm or QEMU to link to a gfx-pps shared library that links to vperfetto. But we want for gfx-pps to be useful also in the non-virtualized case, so there would be some overlap.

Guess we probably want a system mode anyway so we also have ftrace events such as sched_switch, etc ?

Reply:
Sorry, missed that. Added system backend support in be5d4ee21c888039256230ddc3614e0ac6fa2dfa

Turned out that exposed this interesting interaction where we needed to wait a bit for the perfetto service on the host to ack our track event registration; fixed in f4a0859cdc20ccb793719b59c367fe185fb5c316